csrc/abstract_tensor.h

```diff
@@ -253,70 +580,262 @@ struct AbstractTensor {
     return domain.crend();
   }
 
-  AbstractTensor& pushBack(AbstractId id) {
+  AbstractTensorWithInfo<Info>& pushBack(AbstractId id) {
     domain.push_back(std::move(id));
+    info.resize(domain.size());
+    return *this;
+  }
+
+  AbstractTensorWithInfo<Info>& pushBack(AbstractId id, const Info& info) {
+    domain.push_back(std::move(id));
+    info.push_back(info);
     return *this;
   }
 
   template <typename... Args>
-  AbstractTensor& emplaceBack(Args&&... args) {
+  AbstractTensorWithInfo<Info>& emplaceBack(Args&&... args) {
     domain.emplace_back(std::forward<Args>(args)...);
     return *this;
   }
 
   template <typename T>
   bool operator==(T&& t) const {
-    if constexpr (std::is_same_v<AbstractTensor, std::decay_t<T>>) {
-      return domain == t.domain;
-    } else {
-      return domain == std::forward<T>(t);
+    if constexpr (std::is_same_v<
+                      AbstractTensorWithInfo<Info>,
+                      std::decay_t<T>>) {
+      return domain == t.domain && info == t.info;
     }
+    return false;
   }
 
   template <typename T>
   bool operator!=(T&& t) const {
     return !operator==(std::forward<T>(t));
   }
 
-  AbstractTensor& parallelize(int64_t axis, ParallelType parallel_type);
+  AbstractTensorWithInfo<Info>& parallelize(
```

Comment:
nit: I believe `AbstractTensorWithInfo& parallelize(` should be sufficient. C++ will automatically know that `AbstractTensorWithInfo` refers to `AbstractTensorWithInfo<Info>` in this context.